src/state/direct/future.rs

```diff
@@ -43,4 +49,27 @@ where
             delay.await;
         }
     }
+
+    /// Asynchronously resolves as soon as the rate limiter allows it.
+    pub async fn until_n_ready(&self, n: NonZeroU32) -> Result<(), InsufficientCapacity> {
+        self.until_n_ready_with_jitter(n, Jitter::NONE).await
+    }
+
+    /// Asynchronously resolves as soon as the rate limiter allows it, with a randomized wait
+    /// period.
+    pub async fn until_n_ready_with_jitter(
+        &self,
+        n: NonZeroU32,
+        jitter: Jitter,
+    ) -> Result<(), InsufficientCapacity> {
+        match self.check_all(n) {
```

Comment:
Make sure to re-check in a loop here - there's that insidious case where multiple checks are kicked off simultaneously and resolve at the same time. If you don't re-check after the delay has passed, they all go ahead and you've bypassed the rate limit.

Reply:
Right.